### DISCUSSION TITLE
Request for docker images with ELPA or cusolver

### DISCUSSION THREAD

**Discussion by itsuwari** [cp2k/cp2k#3194] [Q&A]
Hello, 

I've been attempting to build CP2K with cuSolverMp on my setup, it's a wild ride without much success. I tried some docker files and cusolvermp install script by @oschuett (huge thanks) but these scripts do not like my setup. It would be wonderful if there are more docker images available. Even builds that aren't fully stable yet would be a huge help. It would make things so much easier for those of us struggling with GPU builds.   

The GPU build available on NGC runs perfectly, and I'm very grateful it's available. The main issue for me is the performance bottleneck with my tiny CPU during diagonalization processes – it's incredibly slow. Meanwhile, I've got several GPUs just waiting to be utilized.

I have tried building with cuda11.8/12.0/12.3, with/without hpcx ucc/ucx, gcc/intel, it seems runs into a common error I just couldn't  get around. its great if some one had similar problems could share some insight. 

```
/usr/bin/ld: /home/b/dt/sw/cp2k/lib/local_cuda/ssmp/libcp2kfm.a(cp_fm_cusolver.o): in function `req_test':
/home/b/dt/sw/cp2k/src/fm/cp_fm_cusolver.c:169:(.text+0x2a2): undefined reference to `MPI_Test'
/usr/bin/ld: /home/b/dt/sw/cp2k/lib/local_cuda/ssmp/libcp2kfm.a(cp_fm_cusolver.o): in function `allgather':
/home/b/dt/sw/cp2k/src/fm/cp_fm_cusolver.c:157:(.text+0x311): undefined reference to `ompi_mpi_byte'
/usr/bin/ld: /home/b/dt/sw/cp2k/src/fm/cp_fm_cusolver.c:157:(.text+0x32a): undefined reference to `MPI_Iallgather'
/usr/bin/ld: /home/b/dt/sw/cp2k/lib/local_cuda/ssmp/libcp2kfm.a(cp_fm_cusolver.o): in function `cp_fm_diag_cusolver':
/home/b/dt/sw/cp2k/src/fm/cp_fm_cusolver.c:197:(.text+0x3e4): undefined reference to `MPI_Comm_f2c'
/usr/bin/ld: /home/b/dt/sw/cp2k/src/fm/cp_fm_cusolver.c:199:(.text+0x3fa): undefined reference to `MPI_Comm_rank'
/usr/bin/ld: /home/b/dt/sw/cp2k/src/fm/cp_fm_cusolver.c:200:(.text+0x40d): undefined reference to `MPI_Comm_size'
/usr/bin/ld: /home/b/dt/sw/cp2k/src/fm/cp_fm_cusolver.c:316:(.text+0x941): undefined reference to `MPI_Barrier'
```

docker version blasts error message  (cusolvermp branch)

```
 SCF WAVEFUNCTION OPTIMIZATION

  Step     Update method      Time    Convergence         Total energy    Change
  ------------------------------------------------------------------------------
error when calling cublasDgemm
error when calling cublasDgemm
error when calling cublasDgemm
error when calling cublasDgemm
error when calling cublasDgemm
error when calling cublasDgemm
error when calling cublasDgemm
error when calling cublasDgemm
error when calling cublasDgemm
error when calling cublasDgemm

```

Best regards.

**Comment by LStuber** [comment]
The latest NGC container (v2023.2) has support for ELPA-GPU, I'd recommend you to give another try. Currently, cusolver / cusolverMP is experimental, so it's probably a bit safer to use ELPA, but hopefully cusolverMp will become default in the future containers.

By the way, I noticed the NGC instructions recommend to use "mpirun --bind-to none" for some reason, which will lead to terrible CPU performance degradation and might be the reason why you thought the CPU diagonalization was so slow. (I've messaged the NGC support to remove this option). The GPU diagonalization will obviously be faster than the CPU version, but for small to medium matrices there is not a huge gap (say <8k).
